File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
Time: O(n)
Space: O(n)
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> result;
        
        if(!root){
            return result;
        }
        
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()){
            TreeNode* rightSide = nullptr;
            int qSize = q.size();
            
            for(int i=0; i<qSize; ++i){
                TreeNode* node = q.front();
                q.pop();
                
                if(node){
                    rightSide = node;
                    if(node->left){
                        q.push(node->left);
                    }
                    if(node->right){
                        q.push(node->right);
                    }
                }
            }
            if(rightSide){
                    result.push_back(rightSide->val);
                }
        }
        return result;
    }
};

/*
int main() {
    // Example usage
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(5);
    root->right->right = new TreeNode(4);

    Solution solution;
    vector<int> result = solution.rightSideView(root);

    // Display the result
    cout << "Right Side View: ";
    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;

    // Remember to free the allocated memory
    delete root->left->right;
    delete root->right->right;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}
*/

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        
        ## intialize a empty result
        res = []
        ## queue initizlized with the root of the tree
        q = collections.deque([root])
        
        ## iterate and pop elements if the queue is not null
        
        while q:
            rightSide = None
            qLength = len(q)
            
            for i in range(qLength):
                node = q.popleft()
                ## check the node should be not null and if not null update the rightSide
                if node:
                    rightSide = node
                    ## append the nodes children
                    q.append(node.left)
                    q.append(node.right)
                    
            if rightSide:
                res.append(rightSide.val)
                
        return res      
        
        